runtime/common/ArgumentConversion.h

```diff
@@ -37,23 +71,39 @@ class ArgumentConverter {
   /// and thereby exclude them from the substitutions.
   void gen_drop_front(const std::vector<void *> &arguments, unsigned numDrop);
 
-  /// Get the list of substitutions that were generated by `gen()`.
-  mlir::SmallVector<cudaq::cc::ArgumentSubstitutionOp> &getSubstitutions() {
-    return substitutions;
+  /// Get the kernel info that were collected by `gen()`.
+  std::list<KernelSubstitutionInfo> &getKernelSubstitutions() {
+    return kernelSubstitutions;
   }
 
-  /// Some substitutions may generate global constant information. Use this
-  /// interface to access both the substitutions and any global constants
-  /// created.
-  mlir::ModuleOp getSubstitutionModule() { return substModule; }
+  bool isRegisteredKernel(const std::string &kernelName) {
+    return std::find(nameRegistry.begin(), nameRegistry.end(), kernelName) !=
+           nameRegistry.end();
+  }
+
+  std::string &registerKernel(const std::string &kernelName) {
+    return nameRegistry.emplace_back(kernelName);
+  }
 
 private:
+  KernelSubstitutionInfo &addKernelInfo(mlir::StringRef kernelName,
+                                        mlir::ModuleOp substModule) {
+    return kernelSubstitutions.emplace_back(kernelName, substModule);
+  }
+
+  /// Memory to store new kernel names generated during argument conversion.
+  /// Use list here to keep references to those elements valid.
+  std::list<std::string> nameRegistry;
+
+  /// Memory to store new kernel info generated during argument conversion.
+  /// Use list here to keep elements sorted in order of creation.
+  std::list<KernelSubstitutionInfo> kernelSubstitutions;
```

Comment:
```suggestion
  SmallVector<KernelSubstitutionInfo> kernelSubstitutions;
```

Reply:
Replaced by `SmallVector<KernelSubstitutionInfo*>` and added a destructor. This allow me to use a small vector and create new kernel infos in order of creation of the corresponding kernels, and avoid references going invalid when the vector is reallocated.